### PR TITLE
Allow Node 22 functions

### LINKED ISSUE #2268: Node 22 Support
### Environment information

```plain text
System:
  OS: macOS 15.1.1
  CPU: (8) arm64 Apple M2
  Memory: 118.66 MB / 16.00 GB
  Shell: /bin/zsh
Binaries:
  Node: 22.9.0 - ~/.nvm/versions/node/v22.9.0/bin/node
  Yarn: undefined - undefined
  npm: 10.8.3 - ~/.nvm/versions/node/v22.9.0/bin/npm
  pnpm: undefined - undefined
```

### Describe the feature

Add support to `@aws-amplify/backend-function` for Node 22, which is a [supported Lambda runtime](https://docs.aws.amazon.com/lambda/latest/dg/lambda-runtimes.html#runtime-deprecation-levels) that was added in [`aws-cdk-lib/aws-lambda` version `2.168.0`](https://github.com/aws/aws-cdk/releases/tag/v2.168.0) on November 20th, 2024

I also suggest dropping support for Node 16 as it is a deprecated runtime.

### Use case

Node 22 is the current LTS version and I would like to use it for my Amplify functions

### PR DESCRIPTION
**Pull request by ataylorme**
## Problem

Closes #2268

## Changes

Add support to `@aws-amplify/backend-function` for Node 22, which is a [supported Lambda runtime](https://docs.aws.amazon.com/lambda/latest/dg/lambda-runtimes.html#runtime-deprecation-levels) that was added in [`aws-cdk-lib/aws-lambda` version `2.168.0`](https://github.com/aws/aws-cdk/releases/tag/v2.168.0) on November 20th, 2024

## Checklist

- [X] If this PR includes a functional change to the runtime behavior of the code, I have added or updated automated test coverage for this change.
- [ ] If this PR requires a change to the [Project Architecture README](../PROJECT_ARCHITECTURE.md), I have included that update in this PR.
- [ ] If this PR requires a docs update, I have linked to that docs PR above.
- [ ] If this PR modifies E2E tests, makes changes to resource provisioning, or makes SDK calls, I have run the PR checks with the `run-e2e` label set.

_By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license._

### CODE REVIEW

File: packages/client-config/API.md [hunk 1]
```diff
@@ -627,7 +627,7 @@ export type CustomClientConfig = {
 export const DEFAULT_CLIENT_CONFIG_VERSION: ClientConfigVersion;
 
 // @public
-export const generateClientConfig: <T extends "1" | "1.1" | "1.2" | "1.3" | "0">(backendIdentifier: DeployedBackendIdentifier, version: T, awsClientProvider?: AWSClientProvider<{
+export const generateClientConfig: <T extends "1.2" | "1.3" | "1.1" | "1" | "0">(backendIdentifier: DeployedBackendIdentifier, version: T, awsClientProvider?: AWSClientProvider<{
```
Comment (sobolk) [inline packages/client-config/API.md:630]: could you please run `npm run clean && npm install && npm run build && npm run update:api` ?
or simply revert this change ?
this is an unwanted diff which sometimes happens in incrementally built local workspace.

File: packages/platform-core/API.md [hunk 2]
```diff
@@ -166,12 +166,12 @@ export const packageJsonSchema: z.ZodObject<{
     type: z.ZodOptional<z.ZodUnion<[z.ZodLiteral<"module">, z.ZodLiteral<"commonjs">]>>;
 }, "strip", z.ZodTypeAny, {
     name?: string | undefined;
-    type?: "module" | "commonjs" | undefined;
     version?: string | undefined;
+    type?: "module" | "commonjs" | undefined;
```
Comment (sobolk) [inline packages/platform-core/API.md:170]: same in this file.

**Review by sobolk — COMMENTED** [review]
LGTM.
I left couple of comments to get checks passing.

File: packages/backend-function/package.json [hunk 3]
```diff
@@ -32,7 +32,7 @@
     "uuid": "^9.0.1"
   },
   "peerDependencies": {
-    "aws-cdk-lib": "^2.158.0",
+    "aws-cdk-lib": "^2.168.0",
```
Comment (sobolk) [inline packages/backend-function/package.json:35]: Please update `aws-cdk-lib` and `aws-cdk` to matching version in all package.json files across the workspace.

Comment (ataylorme) [reply packages/backend-function/package.json:35]: Updated in [d443b1f](https://github.com/aws-amplify/amplify-backend/pull/2269/commits/d443b1f0eaf3bfaf908d38ebf0e33b5434888208)

File: packages/platform-core/API.md (hunk 2 above)
Comment (ataylorme) [reply packages/platform-core/API.md]: Updated in [091f133](https://github.com/aws-amplify/amplify-backend/pull/2269/commits/091f133b18a0bc04843ef8ed6b39d46efe550ce1)

File: packages/client-config/API.md (hunk 1 above)
Comment (ataylorme) [reply packages/client-config/API.md]: Updated in [091f133](https://github.com/aws-amplify/amplify-backend/pull/2269/commits/091f133b18a0bc04843ef8ed6b39d46efe550ce1)

File: packages/backend-function/package.json (hunk 3 above)
Comment (ataylorme) [reply packages/backend-function/package.json:35]: I created a script `update-cdk-in-packages.js` to do this. Sharing in case it is useful to you in the future.

```js
/* eslint-disable */
const rootDir = process.cwd();
import fs from 'fs';
import path from 'path';

async function findPackageJsonFiles(dir) {
    let results = [];
    const list = fs.readdirSync(dir);
    for (let file of list) {
        file = path.resolve(dir, file);
        const stat = fs.statSync(file);
        if (stat && stat.isDirectory()) {
            results = results.concat(await findPackageJsonFiles(file));
        } else if (file.endsWith('package.json')) {
            const relativePath = path.relative(rootDir, file);
            if (/packages\/[\w-]+\/package\.json$/.test(relativePath)) {
                results.push(file);
            }
        }
    }
    return results;
}

function updatePackageVersion(packageJson, packageName, newVersion) {
    if (packageJson.dependencies && packageJson.dependencies[packageName]) {
        packageJson.dependencies[packageName] = newVersion;
    }
    if (packageJson.devDependencies && packageJson.devDependencies[packageName]) {
        packageJson.devDependencies[packageName] = newVersion;
    }
    if (packageJson.peerDependencies && packageJson.peerDependencies[packageName]) {
        packageJson.peerDependencies[packageName] = newVersion;
    }
}

async function updateCdkVersion(filePaths) {
    for (let filePath of filePaths) {
        const packageJson = JSON.parse(fs.readFileSync(filePath, 'utf8'));
        updatePackageVersion(packageJson, 'aws-cdk-lib', '^2.168.0');
        updatePackageVersion(packageJson, 'aws-cdk', '^2.168.0');
        fs.writeFileSync(filePath, JSON.stringify(packageJson, null, 2) + `\n`, 'utf8');
    }
}

async function main() {
    const packageJsonFiles = await findPackageJsonFiles(path.join(rootDir, 'packages'));
    await updateCdkVersion(packageJsonFiles);
    console.log('Updated package.json files:', packageJsonFiles);
}

main();
/* eslint-enable */
```

Comment (sobolk) [reply packages/backend-function/package.json:35]: thank you.
there will be one more thing needed - additional changeset, since most of packages got touched.
Please see https://github.com/aws-amplify/amplify-backend/commit/b56d344e27f48ba235644fb695e6de465028bc65 for reference. This commit contains a changeset file from last time we upgraded CDK.

**Comment by sobolk** [comment]
Looks like some tests are failing. Could you please take a look https://github.com/aws-amplify/amplify-backend/actions/runs/12033789933/job/33549492119?pr=2269 ?

File: packages/backend-function/package.json (hunk 3 above)
Comment (ataylorme) [reply packages/backend-function/package.json:35]: Thanks for the reference, that helps.

File: .changeset/forty-bulldogs-end.md [hunk 4]
```diff
@@ -0,0 +1,7 @@
+---
+'@aws-amplify/backend-function': minor
+---
```
Comment (sobolk) [inline .changeset/forty-bulldogs-end.md:3]:
```suggestion
---
'@aws-amplify/backend-function': minor
'@aws-amplify/backend': minor
---
```

**Review by sobolk — APPROVED** [review]

**Review by vigy02 — APPROVED** [review]

**Comment by sobolk** [comment]
@ataylorme thanks for contributing!
